test/test_algorithms.jl

```diff
@@ -0,0 +1,68 @@
+using Logging
+disable_logging(LogLevel(1))
+
+
+function create_problem(f, range=10)
+    examples = [Herb.HerbData.IOExample(Dict(:x => x), f(x)) for x ∈ 1:range]
+    return Herb.HerbData.Problem(examples, "example"), examples
+end
+
+grammar = @cfgrammar begin
+    C = |(1:5)
+    X = |(1:5)
+    X = C * X
+    X = X + X
+    X = X - X
+    X = X * X
+    X = x
+end
+
+"""
+Expression is an expression like x * x + x * x * x - 5 and max_depth is the max depth
+"""
+macro testmh(expression::String, max_depth=6)
+    return :(
+        @testset "test Metropolis Hastings $($expression)" begin
+        e = Meta.parse("x -> $($expression)")
+        problem, examples = create_problem(eval(e))
+        enumerator = HerbSearch.get_mh_enumerator(grammar, examples, $max_depth, :X, HerbSearch.mean_squared_error)
+        found = Herb.HerbSearch.search_it(grammar, problem, enumerator)
+        println("Wanted $($expression) => Found $found")
+    end
+    )
+end
+
+@testset verbose = true "Algorithms" begin
```

Comment:
I tried hooking this up to the other tests, but there are still a couple of issues:

1. Since you aren't using the `@test` macro, the test overview doesn't show test results. I see that it might be difficult to create `@test`s in for the `@testmh` macro, because you could get equivalent expressions. 
2. Since we are now doing an entire search, we might get an expression that matches the examples but not our original expression, causing flaky tests. 
3. The print statements mess up the test overview.
4. In my opinion, the tests are too complicated for unit tests. It takes 6s to just run the metropolis hastings tests. This is quite long for unit tests.

I think the issues can be fixed by restricting the tests to very simple test cases that test as much of the code as possible. Have a look at e.g. `test_iterators`, where the iterators are tested with very small grammars. Problem 1 & 2 can be solved by evaluating the returned expression on the set of examples.

Reply:
I understand that this tests are not `normal` unit tests because they take long to run.

The main problem I faced with Piotr during development is how to make sure that the random algorithm actually does what it should do and I finds simple expressions like `x * (x + 5) * 2x`. 

It is true that it might find equivalent expressions but it is very hard to enforce a random algorithm to find an exact expression. We just to find something that satisfies the examples provided. 

Sometimes we would introduce bugs and then we would not be able to synthesize expressions that we used to find before.
It is also in a way a nice way to see what are the limits of each algorithm.

I am not sure how to properly implement testing for random algorithms.  We could try to unit test each function from `accept` `propose` etc.

The Simulated Annealing performs pretty bad and does take some time to simulate.

The prints are of course not nice but it shows us what we wanted and what the algorithm provided to inspect.

Reply:
![image](https://user-images.githubusercontent.com/35890341/229475656-1b3e93f4-05f3-48b6-ae22-aaab2a88ce91.png)
When I run `herb_test.jl` in the REPL using VS-Code I see all the test messages.
But I agree that using a test macro like that might not be smart.

Reply:
Eventually we'd like to hook up these tests to CI/CD, meaning that they get automatically executed every time you push or make a pull request. So there should be some tests that either fail or succeed, and they shouldn't take too much time to run.
Currently, the tests show 'None' as a result because they are missing an `@test` in the testsets.
I'm not too familiar with your algorithms, but you mentioned that the result should match all examples, so you could just verify this using `@test`.

The current tests are really nice, but unfortunately, take a long time to run. I'd imagine reducing your search space to only include the building blocks you need might speed things up, e.g.:
```julia
@cfgrammar begin
   X = X * X
   X = X + X
   X = 5
   X = x
end
```
Ideally, this still helps you in finding any bugs. If for some reason that is not the case, we might look into creating a separate test suite or something, but I'm not sure if that is a worthwhile time investment at this point.

Reply:
I don't think the test macro is necessarily a bad idea, just make sure to include an `@test` somewhere.

You can test the expression by evaluating it, i.e. seeing whether it produces 625 as a result.

I wouldn't worry about termination too much, there will probably be a timeout. 
This is a problem that essentially any test has; there can always be an infinite loop somewhere in the code that would make the test not terminate.

Reply:
Testing random algorithms is indeed a bit tricky.
But, you don't have to make them exhaustive.

It would be sufficient to check that a synthesiser, after a certain number of iterations, finds a program that solves more examples than the initial program (either in the number of examples, or reduces mse).

To complement this, you can indeed test the individual accept/propose/neighbourhood functions.